File: language/src/main/java/brj/nodes/DoNode.java
package brj.nodes;

import brj.BridjeLanguage;
import com.oracle.truffle.api.dsl.NodeChild;
import com.oracle.truffle.api.dsl.NodeField;
import com.oracle.truffle.api.dsl.Specialization;
import com.oracle.truffle.api.source.SourceSection;
import org.jetbrains.annotations.NotNull;

@NodeChild(value = "doExprs", type = ExecuteArrayNode.class)
@NodeChild(value = "doExpr", type = ExprNode.class)
@NodeField(name = "sourceSection", type = SourceSection.class)
public abstract class DoNode extends ExprNode {

    protected DoNode(BridjeLanguage lang) {
        super(lang);
    }

    @NotNull
    @Specialization
    public Object execute(Object[] doExprs, Object doExpr) {
        return doExpr;
    }
}

File: language/src/main/kotlin/brj/nodes/DoNode.kt
package brj.nodes

import brj.BridjeNode
import com.oracle.truffle.api.frame.VirtualFrame
import com.oracle.truffle.api.nodes.ExplodeLoop

class DoNode(
    @field:Children private val sideEffectNodes: Array<BridjeNode>,
    @field:Child private var resultNode: BridjeNode
) : BridjeNode() {

    @ExplodeLoop
    override fun execute(frame: VirtualFrame): Any? {
        for (node in sideEffectNodes) {
            node.execute(frame)
        }
        return resultNode.execute(frame)
    }
}
